contractsV5/staking/StakingContract.sol

```diff
@@ -0,0 +1,286 @@
+        if (latestDelegatedAddr == address(0)) {
+            // not delegate to anyone, consider as delegate to yourself
+            latestDelegatedAddress[S] = S;
+            latestDelegatedAddr = S;
+        }
+        uint latestDelegatedStake = latestDelegatedStakes[S];
+        uint latestStakeBal = latestStakeBalance[S];
+
+        if (!hasInited[N][S]) {
+            hasInited[N][S] = true;
+            delegatedAddress[N][S] = latestDelegatedAddr;
+            delegatedStakes[N][S] = latestDelegatedStake;
+            stakes[N][S] = latestStakeBal;
+        }
+
+        if (!hasInited[N + 1][S]) {
+            hasInited[N + 1][S] = true;
+            delegatedAddress[N + 1][S] = latestDelegatedAddr;
+            delegatedStakes[N + 1][S] = latestDelegatedStake;
+            stakes[N + 1][S] = latestStakeBal;
+        }
+    }
+
+    event Delegated(address S, address dAddr, bool isDelegated);
+    function delegate(address dAddr) public returns(bool) {
+        require(dAddr != address(0), "delegate: delegated address should not be 0x0");
+        address S = msg.sender;
+
+        initDataIfNeeded(S);
+
+        uint N = getCurrentEpochNumber();
+
+        address curDelegatedAddr = delegatedAddress[N + 1][S];
+        // nothing changes here
+        if (dAddr == curDelegatedAddr) { return false; }
+
+        uint updatedStake = stakes[N + 1][S];
+
+        // reduce delegatedStakes for curDelegatedAddr if needed
+        if (curDelegatedAddr != S) {
+            initDataIfNeeded(curDelegatedAddr);
+            // by right we don't need to check if delegatedStakes >= stakes
+            require(delegatedStakes[N + 1][curDelegatedAddr] >= updatedStake, "delegate: delegated stake is smaller than next epoch stake");
+            require(latestDelegatedStakes[curDelegatedAddr] >= updatedStake, "delegate: latest delegated stake is smaller than next epoch stake");
+
+            delegatedStakes[N + 1][curDelegatedAddr] -= updatedStake;
+            latestDelegatedStakes[curDelegatedAddr] -= updatedStake;
+
+            emit Delegated(S, curDelegatedAddr, false);
+        }
+
+        latestDelegatedAddress[S] = dAddr;
+        delegatedAddress[N + 1][S] = dAddr;
+
+        // ignore if S delegated back to himself
+        if (dAddr != S) {
+            initDataIfNeeded(dAddr);
+            delegatedStakes[N + 1][dAddr] += updatedStake;
+            latestDelegatedStakes[dAddr] += updatedStake;
+        }
+
+        emit Delegated(S, dAddr, true);
+    }
+
+    event Deposited(uint N, address S, uint amount);
+    function deposit(uint amount) public {
+        require(amount > 0, "deposit: amount to deposit should be positive");
+        // compute epoch number
+        uint N = getCurrentEpochNumber();
+        address S = msg.sender;
+
+        // collect KNC token from sender
+        require(KNC_TOKEN.transferFrom(msg.sender, address(this), amount), "deposit: can not get token");
```

Comment:
using `S` here and using `msg.sender` here, which costs more gas?

Reply:
Updated, should be S instead of msg.sender.